nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+                                continue;
+                            }
+                        }
+
+                        for backfill_tree in backfill_trees {
+                            // Get the tree out of nested structs.
+                            let tree = backfill_tree.unique_tree.tree;
+                            let tree_str = bs58::encode(&tree).into_string();
+                            println!("Backfilling tree: {tree_str}");
+
+                            // Call different methods based on whether tree needs to be backfilled
+                            // completely from seq number 1 or just have any gaps in seq number
+                            // filled.
+                            let result = if backfill_tree.backfill_from_seq_1 {
+                                self.backfill_tree_from_seq_1(&tree).await
+                            } else {
+                                self.fetch_and_plug_gaps(&tree).await
+                            };
+
+                            match result {
+                                Ok(opt_max_seq) => {
+                                    if let Some(max_seq) = opt_max_seq {
+                                        // Debug.
+                                        println!("Successfully backfilled tree: {tree_str}");
+
+                                        // Only delete extra tree rows if fetching and plugging gaps worked.
+                                        if let Err(err) = self
+                                            .delete_extra_rows_and_mark_as_backfilled(
+                                                &tree, max_seq,
+                                            )
+                                            .await
+                                        {
+                                            println!("Error deleting rows and marking as backfilled: {err}");
+                                            self.sleep_and_increase_delay().await;
+                                        } else {
+                                            // Debug.
+                                            println!("Successfully deleted rows up to {max_seq}");
+                                            self.reset_delay()
+                                        }
+                                    } else {
+                                        // Debug.
+                                        println!("Unexpected error, tree was in list, but no rows found for {tree_str}");
+                                        self.sleep_and_increase_delay().await;
+                                    }
+                                }
+                                Err(err) => {
+                                    println!("Failed to fetch and plug gaps for {tree_str}");
+                                    println!("{err}");
+                                    self.sleep_and_increase_delay().await;
```

Comment:
Yes I think this could be a poison pill scenario since it never gives up on a tree.  However, it doesn't seem very "poisonous".   But still you are right, we should plan for it.

If you are interested here's my thoughts below on the main points it could get stuck on a failure, more just my thought process being jotted down:

1. Backfiller gets a list of trees to backfill from the `backfill_items` table, then tries to detect gaps by doing more database queries to that table.  The only failures should be if it can't read the database.  If it gets an empty list it will be happy and move on.  So this is not bad because eventually database read will pass.

2. It gets the relevant slots from RPC provider.  Now if the slot is not available because our slot contains corrupted data, it would keep trying forever (poison).  But if we are asking for a valid slot it will eventually be finalized and be retrieved from the RPC provider.  It's unlikely it will be an invalid slot because we set the slot based on geyser input.  But I won't say never.

3. If the format of the received block data is incorrect, it will keep trying forever (poison).  However it seems unlikely the RPC format would change on us without causing a hard failure in some upgrade/integration testing.  But I won't say never.  We already saw the ReplicaTransaction V2 format show up.

4. Once it has the data its a matter of filtering it on our tree/program IDs, and then serializing which should work because Rust made sure the types are all correct.  But there theoretically could be bs58 decode errors if some of the stuff got corrupted or was encoded wrong originally (poison).  It seems not likely but I won't say never.

5. Once it is serialized it goes to Redis.  If that fails it should not be poison because eventually should be able to put into Redis.  Now if there's something still wrong with the data its a failure on the consumer side.